{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MCB112: Biological Data Analysis \n",
    "## Homework 01: The case of the dead sandmouse"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I've approached the homework questions as follows:\n",
    "\n",
    "1. I converted both files to their own dataframes and processed them in anticipation of questions to answer\n",
    "2. I looked at each file to see their top hits for different parameters relevant to the homework as a first \"look and see\"\n",
    "3. I identified which genes were in one data set and not the other\n",
    "4. I merged the datasets and removed which genes which were not shared in both datasets\n",
    "5. I wrote the merged data set to an excel file and a text file with the requirements from the homework \n",
    "6. I compared the gene expression ratios of each gene to their half-life and solved the case of the dead sandmouse (I think!)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import csv\n",
    "import numpy as np\n",
    "\n",
    "pd.options.display.float_format = '{:.2f}'.format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import text files and convert to DataFrames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>synth_rate</th>\n",
       "      <th>half_life</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZFY</td>\n",
       "      <td>7.20</td>\n",
       "      <td>24.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FAM222B</td>\n",
       "      <td>0.50</td>\n",
       "      <td>15.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MRPS33</td>\n",
       "      <td>0.90</td>\n",
       "      <td>25.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LSM1</td>\n",
       "      <td>2.50</td>\n",
       "      <td>17.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TIAL1</td>\n",
       "      <td>31.20</td>\n",
       "      <td>6.80</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name  synth_rate  half_life\n",
       "1       ZFY        7.20      24.30\n",
       "2   FAM222B        0.50      15.40\n",
       "3    MRPS33        0.90      25.40\n",
       "4      LSM1        2.50      17.40\n",
       "5     TIAL1       31.20       6.80"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#open Adler text file and convert to pandas DataFrame object\n",
    "\n",
    "with open(\"/Volumes/Macintosh HD/Users/Hailey/Dropbox/MCB112/Homework_1/Adler_SuppTable2.txt\") as Adler:\n",
    "    Adler_txt = Adler.read()\n",
    "    Adler_list = Adler_txt.split(\"\\n\")\n",
    "    Adler_series = pd.Series(Adler_list)\n",
    "    Adler_df = Adler_series.str.split(expand=True)\n",
    "Adler_df.columns = ['gene_name', 'synth_rate', 'half_life', 'todelete']\n",
    "Adler_df = Adler_df.drop(0)\n",
    "Adler_df = Adler_df.drop('todelete', axis=1)\n",
    "Adler_df['half_life'] = pd.to_numeric(Adler_df['half_life'])\n",
    "Adler_df['synth_rate'] = pd.to_numeric(Adler_df['synth_rate'])\n",
    "\n",
    "Adler_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>0h</th>\n",
       "      <th>12h</th>\n",
       "      <th>24h</th>\n",
       "      <th>48h</th>\n",
       "      <th>96h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>anise</td>\n",
       "      <td>71.10</td>\n",
       "      <td>36.50</td>\n",
       "      <td>15.60</td>\n",
       "      <td>3.10</td>\n",
       "      <td>0.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>apricot</td>\n",
       "      <td>28.20</td>\n",
       "      <td>31.00</td>\n",
       "      <td>36.10</td>\n",
       "      <td>27.60</td>\n",
       "      <td>11.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>artichoke</td>\n",
       "      <td>75.30</td>\n",
       "      <td>73.20</td>\n",
       "      <td>61.00</td>\n",
       "      <td>42.40</td>\n",
       "      <td>9.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>arugula</td>\n",
       "      <td>50.70</td>\n",
       "      <td>46.30</td>\n",
       "      <td>43.50</td>\n",
       "      <td>25.30</td>\n",
       "      <td>4.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>asparagus</td>\n",
       "      <td>17.30</td>\n",
       "      <td>4.50</td>\n",
       "      <td>1.10</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_name    0h   12h   24h   48h   96h\n",
       "1      anise 71.10 36.50 15.60  3.10  0.10\n",
       "2    apricot 28.20 31.00 36.10 27.60 11.90\n",
       "3  artichoke 75.30 73.20 61.00 42.40  9.70\n",
       "4    arugula 50.70 46.30 43.50 25.30  4.70\n",
       "5  asparagus 17.30  4.50  1.10  0.10  0.00"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#open Moriarty text file and convert to pandas DataFrame object\n",
    "\n",
    "with open(\"/Volumes/Macintosh HD/Users/Hailey/Dropbox/MCB112/Homework_1/Moriarty_SuppTable1.txt\") as Moriarty:\n",
    "    Moriarty_txt = Moriarty.read()\n",
    "    Moriarty_list = Moriarty_txt.split(\"\\n\")\n",
    "    Moriarty_series = pd.Series(Moriarty_list)\n",
    "    Moriarty_df = Moriarty_series.str.split(expand=True)\n",
    "Moriarty_df.columns = ['gene_name', '0h', '12h', '24h', '48h', '96h', 'todelete']\n",
    "Moriarty_df = Moriarty_df.drop(0)\n",
    "Moriarty_df = Moriarty_df.drop('todelete', axis=1)\n",
    "Moriarty_df['0h'] = pd.to_numeric(Moriarty_df['0h'])\n",
    "Moriarty_df['12h'] = pd.to_numeric(Moriarty_df['12h'])\n",
    "Moriarty_df['24h'] = pd.to_numeric(Moriarty_df['24h'])\n",
    "Moriarty_df['48h'] = pd.to_numeric(Moriarty_df['48h'])\n",
    "Moriarty_df['96h'] = pd.to_numeric(Moriarty_df['96h'])\n",
    "\n",
    "Moriarty_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Take a look at top genes for different parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#preemptive merging to look at all data for top hits depending on the parameter we are looking at\n",
    "all_genes_df = pd.merge(Adler_df, Moriarty_df, on=['gene_name'], how='outer')\n",
    "\n",
    "#check out top hits for the half-life, synthesis rates, and expression levels at 0 and 96 hrs from both datasets\n",
    "top_ten_half_life = all_genes_df.nlargest(10, 'half_life')\n",
    "top_ten_synth_rate = all_genes_df.nlargest(10, 'synth_rate')\n",
    "top_ten_at_t0 = all_genes_df.nlargest(10, \"0h\")\n",
    "top_ten_at_t96 = all_genes_df.nlargest(10, \"96h\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>synth_rate</th>\n",
       "      <th>half_life</th>\n",
       "      <th>0h</th>\n",
       "      <th>12h</th>\n",
       "      <th>24h</th>\n",
       "      <th>48h</th>\n",
       "      <th>96h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>975</th>\n",
       "      <td>BEX2</td>\n",
       "      <td>3.60</td>\n",
       "      <td>88.70</td>\n",
       "      <td>329.30</td>\n",
       "      <td>600.20</td>\n",
       "      <td>1292.20</td>\n",
       "      <td>3126.90</td>\n",
       "      <td>13026.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8993</th>\n",
       "      <td>LIPH</td>\n",
       "      <td>7.90</td>\n",
       "      <td>63.40</td>\n",
       "      <td>446.80</td>\n",
       "      <td>1071.20</td>\n",
       "      <td>1748.40</td>\n",
       "      <td>3692.40</td>\n",
       "      <td>12674.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6029</th>\n",
       "      <td>NLK</td>\n",
       "      <td>4.10</td>\n",
       "      <td>62.90</td>\n",
       "      <td>264.90</td>\n",
       "      <td>492.40</td>\n",
       "      <td>945.40</td>\n",
       "      <td>2000.60</td>\n",
       "      <td>7174.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1174</th>\n",
       "      <td>U2AF1</td>\n",
       "      <td>9.60</td>\n",
       "      <td>56.40</td>\n",
       "      <td>571.10</td>\n",
       "      <td>1015.00</td>\n",
       "      <td>1810.90</td>\n",
       "      <td>3913.00</td>\n",
       "      <td>11651.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3500</th>\n",
       "      <td>TMEM186</td>\n",
       "      <td>21.70</td>\n",
       "      <td>53.90</td>\n",
       "      <td>1285.00</td>\n",
       "      <td>2154.70</td>\n",
       "      <td>4079.00</td>\n",
       "      <td>8581.40</td>\n",
       "      <td>25213.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_name  synth_rate  half_life      0h     12h     24h     48h      96h\n",
       "975       BEX2        3.60      88.70  329.30  600.20 1292.20 3126.90 13026.80\n",
       "8993      LIPH        7.90      63.40  446.80 1071.20 1748.40 3692.40 12674.20\n",
       "6029       NLK        4.10      62.90  264.90  492.40  945.40 2000.60  7174.30\n",
       "1174     U2AF1        9.60      56.40  571.10 1015.00 1810.90 3913.00 11651.90\n",
       "3500   TMEM186       21.70      53.90 1285.00 2154.70 4079.00 8581.40 25213.00"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_ten_half_life.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>synth_rate</th>\n",
       "      <th>half_life</th>\n",
       "      <th>0h</th>\n",
       "      <th>12h</th>\n",
       "      <th>24h</th>\n",
       "      <th>48h</th>\n",
       "      <th>96h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9622</th>\n",
       "      <td>LINC00176</td>\n",
       "      <td>118.60</td>\n",
       "      <td>13.80</td>\n",
       "      <td>1615.00</td>\n",
       "      <td>1558.60</td>\n",
       "      <td>1488.10</td>\n",
       "      <td>861.40</td>\n",
       "      <td>185.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>889</th>\n",
       "      <td>MAP3K9</td>\n",
       "      <td>102.80</td>\n",
       "      <td>14.20</td>\n",
       "      <td>1489.00</td>\n",
       "      <td>1509.30</td>\n",
       "      <td>1412.00</td>\n",
       "      <td>963.70</td>\n",
       "      <td>217.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5349</th>\n",
       "      <td>FTSJ1</td>\n",
       "      <td>99.50</td>\n",
       "      <td>15.60</td>\n",
       "      <td>1582.10</td>\n",
       "      <td>1794.50</td>\n",
       "      <td>1553.10</td>\n",
       "      <td>1141.60</td>\n",
       "      <td>417.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19090</th>\n",
       "      <td>PIGP</td>\n",
       "      <td>83.30</td>\n",
       "      <td>17.70</td>\n",
       "      <td>1501.10</td>\n",
       "      <td>1865.10</td>\n",
       "      <td>1933.20</td>\n",
       "      <td>1678.40</td>\n",
       "      <td>801.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15314</th>\n",
       "      <td>GDAP2</td>\n",
       "      <td>73.90</td>\n",
       "      <td>9.40</td>\n",
       "      <td>685.80</td>\n",
       "      <td>477.20</td>\n",
       "      <td>280.20</td>\n",
       "      <td>70.40</td>\n",
       "      <td>3.10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_name  synth_rate  half_life      0h     12h     24h     48h    96h\n",
       "9622   LINC00176      118.60      13.80 1615.00 1558.60 1488.10  861.40 185.10\n",
       "889       MAP3K9      102.80      14.20 1489.00 1509.30 1412.00  963.70 217.70\n",
       "5349       FTSJ1       99.50      15.60 1582.10 1794.50 1553.10 1141.60 417.80\n",
       "19090       PIGP       83.30      17.70 1501.10 1865.10 1933.20 1678.40 801.90\n",
       "15314      GDAP2       73.90       9.40  685.80  477.20  280.20   70.40   3.10"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_ten_synth_rate.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>synth_rate</th>\n",
       "      <th>half_life</th>\n",
       "      <th>0h</th>\n",
       "      <th>12h</th>\n",
       "      <th>24h</th>\n",
       "      <th>48h</th>\n",
       "      <th>96h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12276</th>\n",
       "      <td>ATP6V1D</td>\n",
       "      <td>64.40</td>\n",
       "      <td>26.10</td>\n",
       "      <td>1655.80</td>\n",
       "      <td>2452.90</td>\n",
       "      <td>3396.00</td>\n",
       "      <td>4211.40</td>\n",
       "      <td>5210.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9622</th>\n",
       "      <td>LINC00176</td>\n",
       "      <td>118.60</td>\n",
       "      <td>13.80</td>\n",
       "      <td>1615.00</td>\n",
       "      <td>1558.60</td>\n",
       "      <td>1488.10</td>\n",
       "      <td>861.40</td>\n",
       "      <td>185.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5349</th>\n",
       "      <td>FTSJ1</td>\n",
       "      <td>99.50</td>\n",
       "      <td>15.60</td>\n",
       "      <td>1582.10</td>\n",
       "      <td>1794.50</td>\n",
       "      <td>1553.10</td>\n",
       "      <td>1141.60</td>\n",
       "      <td>417.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6865</th>\n",
       "      <td>LAMP3</td>\n",
       "      <td>48.80</td>\n",
       "      <td>30.90</td>\n",
       "      <td>1564.10</td>\n",
       "      <td>2399.70</td>\n",
       "      <td>3202.60</td>\n",
       "      <td>5083.10</td>\n",
       "      <td>8526.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19090</th>\n",
       "      <td>PIGP</td>\n",
       "      <td>83.30</td>\n",
       "      <td>17.70</td>\n",
       "      <td>1501.10</td>\n",
       "      <td>1865.10</td>\n",
       "      <td>1933.20</td>\n",
       "      <td>1678.40</td>\n",
       "      <td>801.90</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_name  synth_rate  half_life      0h     12h     24h     48h  \\\n",
       "12276    ATP6V1D       64.40      26.10 1655.80 2452.90 3396.00 4211.40   \n",
       "9622   LINC00176      118.60      13.80 1615.00 1558.60 1488.10  861.40   \n",
       "5349       FTSJ1       99.50      15.60 1582.10 1794.50 1553.10 1141.60   \n",
       "6865       LAMP3       48.80      30.90 1564.10 2399.70 3202.60 5083.10   \n",
       "19090       PIGP       83.30      17.70 1501.10 1865.10 1933.20 1678.40   \n",
       "\n",
       "          96h  \n",
       "12276 5210.00  \n",
       "9622   185.10  \n",
       "5349   417.80  \n",
       "6865  8526.40  \n",
       "19090  801.90  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_ten_at_t0.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>synth_rate</th>\n",
       "      <th>half_life</th>\n",
       "      <th>0h</th>\n",
       "      <th>12h</th>\n",
       "      <th>24h</th>\n",
       "      <th>48h</th>\n",
       "      <th>96h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3500</th>\n",
       "      <td>TMEM186</td>\n",
       "      <td>21.70</td>\n",
       "      <td>53.90</td>\n",
       "      <td>1285.00</td>\n",
       "      <td>2154.70</td>\n",
       "      <td>4079.00</td>\n",
       "      <td>8581.40</td>\n",
       "      <td>25213.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17808</th>\n",
       "      <td>CAND1</td>\n",
       "      <td>21.40</td>\n",
       "      <td>45.30</td>\n",
       "      <td>1039.70</td>\n",
       "      <td>1815.30</td>\n",
       "      <td>2615.80</td>\n",
       "      <td>5888.80</td>\n",
       "      <td>13543.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>975</th>\n",
       "      <td>BEX2</td>\n",
       "      <td>3.60</td>\n",
       "      <td>88.70</td>\n",
       "      <td>329.30</td>\n",
       "      <td>600.20</td>\n",
       "      <td>1292.20</td>\n",
       "      <td>3126.90</td>\n",
       "      <td>13026.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8993</th>\n",
       "      <td>LIPH</td>\n",
       "      <td>7.90</td>\n",
       "      <td>63.40</td>\n",
       "      <td>446.80</td>\n",
       "      <td>1071.20</td>\n",
       "      <td>1748.40</td>\n",
       "      <td>3692.40</td>\n",
       "      <td>12674.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1259</th>\n",
       "      <td>GTDC1</td>\n",
       "      <td>21.10</td>\n",
       "      <td>44.20</td>\n",
       "      <td>960.30</td>\n",
       "      <td>1724.10</td>\n",
       "      <td>2769.70</td>\n",
       "      <td>5484.00</td>\n",
       "      <td>12031.20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name  synth_rate  half_life      0h     12h     24h     48h  \\\n",
       "3500    TMEM186       21.70      53.90 1285.00 2154.70 4079.00 8581.40   \n",
       "17808     CAND1       21.40      45.30 1039.70 1815.30 2615.80 5888.80   \n",
       "975        BEX2        3.60      88.70  329.30  600.20 1292.20 3126.90   \n",
       "8993       LIPH        7.90      63.40  446.80 1071.20 1748.40 3692.40   \n",
       "1259      GTDC1       21.10      44.20  960.30 1724.10 2769.70 5484.00   \n",
       "\n",
       "           96h  \n",
       "3500  25213.00  \n",
       "17808 13543.70  \n",
       "975   13026.80  \n",
       "8993  12674.20  \n",
       "1259  12031.20  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_ten_at_t96.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Remove the genes that aren't in the other dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20031      None\n",
      "20032    15-Sep\n",
      "20033     2-Mar\n",
      "20034     2-Mar\n",
      "20035     1-Mar\n",
      "20036     1-Mar\n",
      "20037    10-Sep\n",
      "20038     7-Mar\n",
      "20039     4-Mar\n",
      "20040     2-Sep\n",
      "20041    11-Sep\n",
      "20042     6-Mar\n",
      "20043    11-Mar\n",
      "20044     3-Mar\n",
      "20045     8-Sep\n",
      "20046     7-Sep\n",
      "20047    14-Sep\n",
      "20048     6-Sep\n",
      "20049     1-Dec\n",
      "20050     8-Mar\n",
      "20051     5-Mar\n",
      "20052     9-Mar\n",
      "20053    12-Sep\n",
      "20054     1-Sep\n",
      "20055     4-Sep\n",
      "20056    10-Mar\n",
      "20057     9-Sep\n",
      "20058     5-Sep\n",
      "20059     3-Sep\n",
      "Name: gene_name, dtype: object\n",
      "583       MARCH4\n",
      "1360      MARCH1\n",
      "3137      SEPT10\n",
      "3963      MARCH8\n",
      "3988       SEP15\n",
      "5009      MARCH2\n",
      "5440      SEPT14\n",
      "6195      MARCH6\n",
      "7003       SEPT5\n",
      "7195       SEPT9\n",
      "7205     MARCH11\n",
      "8126       SEPT3\n",
      "8598       SEPT1\n",
      "9149      MARCH9\n",
      "9233      MARCH3\n",
      "9767       MARC2\n",
      "11389      SEPT8\n",
      "11598     MARCH7\n",
      "12927      SEPT7\n",
      "14376     MARCH5\n",
      "14749      MARC1\n",
      "16210      SEPT4\n",
      "18023     SEPT11\n",
      "18053    MARCH10\n",
      "18248      SEPT6\n",
      "19214       DEC1\n",
      "19633     SEPT12\n",
      "19892      SEPT2\n",
      "20031       None\n",
      "Name: gene_name, dtype: object\n"
     ]
    }
   ],
   "source": [
    "#basic approach to removing the data that is missing in one dataframe from the other is to find those genes for which \n",
    "#the information in NaN --> NaN in synth_rate indicates presence in Moriarty and not Adler, and vice versa\n",
    "#example below\n",
    "all_genes_df_0 = pd.merge(Adler_df, Moriarty_df, on=['gene_name'], how='outer')\n",
    "\n",
    "in_Adler_only = all_genes_df_0.loc[all_genes_df_0['0h'].isna()]\n",
    "in_Moriarty_only = all_genes_df_0.loc[all_genes_df_0['synth_rate'].isna()]\n",
    "\n",
    "print(in_Moriarty_only['gene_name'])\n",
    "print(in_Adler_only['gene_name'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Most genes are shared between Moriarty and Adler datasets, but both have corrupted gene names which have been converted to dates and should no longer be included in the dataset. There could be an exception with Marc1 and Marc2 which may not have been corrupted, and are just not included in the other dataset. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Merge datasets, get ratios of expression levels, remove all uncommon genes, and output file with parameters specified by homework"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/Hailey/anaconda/envs/python3/lib/python3.5/site-packages/ipykernel/__main__.py:19: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/Users/Hailey/anaconda/envs/python3/lib/python3.5/site-packages/ipykernel/__main__.py:20: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/Users/Hailey/anaconda/envs/python3/lib/python3.5/site-packages/ipykernel/__main__.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/Users/Hailey/anaconda/envs/python3/lib/python3.5/site-packages/ipykernel/__main__.py:22: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "all_genes_df_0 = pd.merge(Adler_df, Moriarty_df, on=['gene_name'], how='outer')\n",
    "\n",
    "#double-checking that we are actually removing all possible empty/use-less entries -- excel errors from each file and\n",
    "#'none' rows\n",
    "\n",
    "#excel_error_df1 = all_genes_df_0.loc[all_genes_df_0['0h'].isna()]\n",
    "#excel_error_df2 = all_genes_df_0.loc[all_genes_df_0['12h'].isna()]\n",
    "#excel_error_df3 = all_genes_df_0.loc[all_genes_df_0['24h'].isna()]\n",
    "#excel_error_df4 = all_genes_df_0.loc[all_genes_df_0['48h'].isna()]\n",
    "#excel_error_df5 = all_genes_df_0.loc[all_genes_df_0['96h'].isna()]\n",
    "#excel_error_df6 = all_genes_df_0.loc[all_genes_df_0['synth_rate'].isna()]\n",
    "#excel_error_df7 = all_genes_df_0.loc[all_genes_df_0['half_life'].isna()]\n",
    "#print(len(excel_error_df1), len(excel_error_df2), len(excel_error_df3), len(excel_error_df4), len(excel_error_df5), len(excel_error_df6), len(excel_error_df7)) \n",
    "\n",
    "#remove all genes that don't match in each dataframe\n",
    "all_genes_df_1 = all_genes_df_0.dropna()\n",
    "\n",
    "#normalize the expression levels of each time point by the initial timepoint. Normalize the initial timepoint last.\n",
    "all_genes_df_1['tpm[12h]/tpm[0]'] = all_genes_df_1['12h'].div(all_genes_df_1['0h'])\n",
    "all_genes_df_1['tpm[24h]/tpm[0]'] = all_genes_df_1['24h'].div(all_genes_df_1['0h'])\n",
    "all_genes_df_1['tpm[48h]/tpm[0]'] = all_genes_df_1['48h']/all_genes_df_1['0h']\n",
    "all_genes_df_1['tpm[96h]/tpm[0]'] = all_genes_df_1['96h']/all_genes_df_1['0h']\n",
    "\n",
    "#remove the original info from Moriarty's set to fit format for homework assignment\n",
    "all_genes_df_1 = all_genes_df_1.drop(['0h', '12h', '24h', '48h', '96h'], axis=1)\n",
    "all_genes_df_1 = all_genes_df_1.round(2)\n",
    "\n",
    "#write normalized, merged dataset to an excel file, and remove index \n",
    "excel_file_writer = pd.ExcelWriter(\"/Volumes/Macintosh HD/Users/Hailey/Dropbox/MCB112/Homework_1/normalized_cleaned_merged_dataset.xlsx\")\n",
    "all_genes_df_1.to_excel(excel_file_writer,'Adler_Moriarty_normalized', index=False)\n",
    "\n",
    "excel_file_writer.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#make text file list to appease the biological data gods.\n",
    "A_M_list = all_genes_df_1.values.tolist()\n",
    "\n",
    "with open('/Volumes/Macintosh HD/Users/Hailey/Dropbox/MCB112/Homework_1/normalized_cleaned_merged_dataset.txt', 'w') as f:\n",
    "    f.write('%-20s%-20s%-20s%-20s%-20s%-20s%-20s\\n'%('Gene name', 'Synthesis rate', 'Half-life', 'tpm[12h]/tpm[0h]', 'tpm[24h]/tpm[0h]', 'tpm[48h]/tpm[0h]', 'tpm[96h]/tpm[0h]'))\n",
    "    for line in A_M_list:\n",
    "        for value in line:\n",
    "            f.write('%-20s'.format() % value)\n",
    "        f.write('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>synth_rate</th>\n",
       "      <th>half_life</th>\n",
       "      <th>tpm[12h]/tpm[0]</th>\n",
       "      <th>tpm[24h]/tpm[0]</th>\n",
       "      <th>tpm[48h]/tpm[0]</th>\n",
       "      <th>tpm[96h]/tpm[0]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZFY</td>\n",
       "      <td>7.20</td>\n",
       "      <td>24.30</td>\n",
       "      <td>1.47</td>\n",
       "      <td>1.88</td>\n",
       "      <td>2.22</td>\n",
       "      <td>2.35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FAM222B</td>\n",
       "      <td>0.50</td>\n",
       "      <td>15.40</td>\n",
       "      <td>1.08</td>\n",
       "      <td>0.96</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MRPS33</td>\n",
       "      <td>0.90</td>\n",
       "      <td>25.40</td>\n",
       "      <td>1.48</td>\n",
       "      <td>1.81</td>\n",
       "      <td>2.61</td>\n",
       "      <td>2.62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LSM1</td>\n",
       "      <td>2.50</td>\n",
       "      <td>17.40</td>\n",
       "      <td>1.11</td>\n",
       "      <td>1.13</td>\n",
       "      <td>1.06</td>\n",
       "      <td>0.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TIAL1</td>\n",
       "      <td>31.20</td>\n",
       "      <td>6.80</td>\n",
       "      <td>0.39</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name  synth_rate  half_life  tpm[12h]/tpm[0]  tpm[24h]/tpm[0]  \\\n",
       "0       ZFY        7.20      24.30             1.47             1.88   \n",
       "1   FAM222B        0.50      15.40             1.08             0.96   \n",
       "2    MRPS33        0.90      25.40             1.48             1.81   \n",
       "3      LSM1        2.50      17.40             1.11             1.13   \n",
       "4     TIAL1       31.20       6.80             0.39             0.15   \n",
       "\n",
       "   tpm[48h]/tpm[0]  tpm[96h]/tpm[0]  \n",
       "0             2.22             2.35  \n",
       "1             0.68             0.22  \n",
       "2             2.61             2.62  \n",
       "3             1.06             0.46  \n",
       "4             0.02             0.00  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_genes_df_1.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Checking for top genes after normalization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "top_ten_half_life_n = all_genes_df_1.nlargest(10, 'half_life')\n",
    "top_ten_synth_rate_n = all_genes_df_1.nlargest(10, 'synth_rate')\n",
    "top_ten_at_t96_n = all_genes_df_1.nlargest(10, 'tpm[96h]/tpm[0]')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>synth_rate</th>\n",
       "      <th>half_life</th>\n",
       "      <th>tpm[12h]/tpm[0]</th>\n",
       "      <th>tpm[24h]/tpm[0]</th>\n",
       "      <th>tpm[48h]/tpm[0]</th>\n",
       "      <th>tpm[96h]/tpm[0]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>975</th>\n",
       "      <td>BEX2</td>\n",
       "      <td>3.60</td>\n",
       "      <td>88.70</td>\n",
       "      <td>1.82</td>\n",
       "      <td>3.92</td>\n",
       "      <td>9.50</td>\n",
       "      <td>39.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8993</th>\n",
       "      <td>LIPH</td>\n",
       "      <td>7.90</td>\n",
       "      <td>63.40</td>\n",
       "      <td>2.40</td>\n",
       "      <td>3.91</td>\n",
       "      <td>8.26</td>\n",
       "      <td>28.37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6029</th>\n",
       "      <td>NLK</td>\n",
       "      <td>4.10</td>\n",
       "      <td>62.90</td>\n",
       "      <td>1.86</td>\n",
       "      <td>3.57</td>\n",
       "      <td>7.55</td>\n",
       "      <td>27.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1174</th>\n",
       "      <td>U2AF1</td>\n",
       "      <td>9.60</td>\n",
       "      <td>56.40</td>\n",
       "      <td>1.78</td>\n",
       "      <td>3.17</td>\n",
       "      <td>6.85</td>\n",
       "      <td>20.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3500</th>\n",
       "      <td>TMEM186</td>\n",
       "      <td>21.70</td>\n",
       "      <td>53.90</td>\n",
       "      <td>1.68</td>\n",
       "      <td>3.17</td>\n",
       "      <td>6.68</td>\n",
       "      <td>19.62</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_name  synth_rate  half_life  tpm[12h]/tpm[0]  tpm[24h]/tpm[0]  \\\n",
       "975       BEX2        3.60      88.70             1.82             3.92   \n",
       "8993      LIPH        7.90      63.40             2.40             3.91   \n",
       "6029       NLK        4.10      62.90             1.86             3.57   \n",
       "1174     U2AF1        9.60      56.40             1.78             3.17   \n",
       "3500   TMEM186       21.70      53.90             1.68             3.17   \n",
       "\n",
       "      tpm[48h]/tpm[0]  tpm[96h]/tpm[0]  \n",
       "975              9.50            39.56  \n",
       "8993             8.26            28.37  \n",
       "6029             7.55            27.08  \n",
       "1174             6.85            20.40  \n",
       "3500             6.68            19.62  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_ten_half_life_n.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>synth_rate</th>\n",
       "      <th>half_life</th>\n",
       "      <th>tpm[12h]/tpm[0]</th>\n",
       "      <th>tpm[24h]/tpm[0]</th>\n",
       "      <th>tpm[48h]/tpm[0]</th>\n",
       "      <th>tpm[96h]/tpm[0]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9622</th>\n",
       "      <td>LINC00176</td>\n",
       "      <td>118.60</td>\n",
       "      <td>13.80</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.92</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>889</th>\n",
       "      <td>MAP3K9</td>\n",
       "      <td>102.80</td>\n",
       "      <td>14.20</td>\n",
       "      <td>1.01</td>\n",
       "      <td>0.95</td>\n",
       "      <td>0.65</td>\n",
       "      <td>0.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5349</th>\n",
       "      <td>FTSJ1</td>\n",
       "      <td>99.50</td>\n",
       "      <td>15.60</td>\n",
       "      <td>1.13</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.72</td>\n",
       "      <td>0.26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19090</th>\n",
       "      <td>PIGP</td>\n",
       "      <td>83.30</td>\n",
       "      <td>17.70</td>\n",
       "      <td>1.24</td>\n",
       "      <td>1.29</td>\n",
       "      <td>1.12</td>\n",
       "      <td>0.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15314</th>\n",
       "      <td>GDAP2</td>\n",
       "      <td>73.90</td>\n",
       "      <td>9.40</td>\n",
       "      <td>0.70</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_name  synth_rate  half_life  tpm[12h]/tpm[0]  tpm[24h]/tpm[0]  \\\n",
       "9622   LINC00176      118.60      13.80             0.97             0.92   \n",
       "889       MAP3K9      102.80      14.20             1.01             0.95   \n",
       "5349       FTSJ1       99.50      15.60             1.13             0.98   \n",
       "19090       PIGP       83.30      17.70             1.24             1.29   \n",
       "15314      GDAP2       73.90       9.40             0.70             0.41   \n",
       "\n",
       "       tpm[48h]/tpm[0]  tpm[96h]/tpm[0]  \n",
       "9622              0.53             0.11  \n",
       "889               0.65             0.15  \n",
       "5349              0.72             0.26  \n",
       "19090             1.12             0.53  \n",
       "15314             0.10             0.00  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_ten_synth_rate_n.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>synth_rate</th>\n",
       "      <th>half_life</th>\n",
       "      <th>tpm[12h]/tpm[0]</th>\n",
       "      <th>tpm[24h]/tpm[0]</th>\n",
       "      <th>tpm[48h]/tpm[0]</th>\n",
       "      <th>tpm[96h]/tpm[0]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>975</th>\n",
       "      <td>BEX2</td>\n",
       "      <td>3.60</td>\n",
       "      <td>88.70</td>\n",
       "      <td>1.82</td>\n",
       "      <td>3.92</td>\n",
       "      <td>9.50</td>\n",
       "      <td>39.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8993</th>\n",
       "      <td>LIPH</td>\n",
       "      <td>7.90</td>\n",
       "      <td>63.40</td>\n",
       "      <td>2.40</td>\n",
       "      <td>3.91</td>\n",
       "      <td>8.26</td>\n",
       "      <td>28.37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6029</th>\n",
       "      <td>NLK</td>\n",
       "      <td>4.10</td>\n",
       "      <td>62.90</td>\n",
       "      <td>1.86</td>\n",
       "      <td>3.57</td>\n",
       "      <td>7.55</td>\n",
       "      <td>27.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1174</th>\n",
       "      <td>U2AF1</td>\n",
       "      <td>9.60</td>\n",
       "      <td>56.40</td>\n",
       "      <td>1.78</td>\n",
       "      <td>3.17</td>\n",
       "      <td>6.85</td>\n",
       "      <td>20.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3500</th>\n",
       "      <td>TMEM186</td>\n",
       "      <td>21.70</td>\n",
       "      <td>53.90</td>\n",
       "      <td>1.68</td>\n",
       "      <td>3.17</td>\n",
       "      <td>6.68</td>\n",
       "      <td>19.62</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_name  synth_rate  half_life  tpm[12h]/tpm[0]  tpm[24h]/tpm[0]  \\\n",
       "975       BEX2        3.60      88.70             1.82             3.92   \n",
       "8993      LIPH        7.90      63.40             2.40             3.91   \n",
       "6029       NLK        4.10      62.90             1.86             3.57   \n",
       "1174     U2AF1        9.60      56.40             1.78             3.17   \n",
       "3500   TMEM186       21.70      53.90             1.68             3.17   \n",
       "\n",
       "      tpm[48h]/tpm[0]  tpm[96h]/tpm[0]  \n",
       "975              9.50            39.56  \n",
       "8993             8.26            28.37  \n",
       "6029             7.55            27.08  \n",
       "1174             6.85            20.40  \n",
       "3500             6.68            19.62  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_ten_at_t96_n.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Answer: The sand mouse experiment did not account for the initial expression levels at the time of death. It turns out that the genes with the highest level fo expression at 96 hours are also the same genes that have the longest half-life. This is reasonable, since the genes with the highest expression at the longest time point will be the transcripts that are most resistant to degradation and should remain at a relatively higher level at 96 hours post-mortem compared to other transcripts. "
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda env:python3]",
   "language": "python",
   "name": "conda-env-python3-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
